socks5.c

```diff
@@ -0,0 +1,337 @@
+        conn[3] = 0x04; // ATYP = IPV6
+        memcpy((void *)(conn + 4), (void *)&ina6, 16);
+        conn[20] = port >> 8;
+        conn[21] = port & 0xFF;
+        conn_len = 22;
+   } else {
+        // assume a resolvable domain name
+        // previously validate to be 0 < x <= 255 bytes
+        uint8_t alen = (uint8_t)strlen(addr);
+        conn[3] = 0x03; // ATYP = DOMAINNAME
+        conn[4] = alen;
+        memcpy((void *)(conn+ 5), (void *)addr, alen);
+        conn[5+alen] = port >> 8;
+        conn[6+alen] = port & 0xFF;
+        conn_len = 7 + alen;
+    }
+    int err = dill_bsend(obj->u, conn, conn_len, deadline);
+    if (dill_slow(err)) {return -1;}
+    // reply has the same form/length limit, reuse request as reply buffer
+    // read first 5 octets (through ATYP) to determine how much more to read
+    err = dill_brecv(obj->u, conn, 5, deadline);
+    if (dill_slow(err)) {return -1;}
+    switch (conn[3]) {
+        case 0x01:
+            // IPv4, already read 1 bytes, read remaining 3 and port
+            err = dill_brecv(obj->u, conn + 5, 5, deadline);
+            if (dill_slow(err)) {return -1;}
+            break;
+        case 0x03:
+            // IPv4, already read 1 bytes, read remaining 15 and port
+            err = dill_brecv(obj->u, conn + 5, 17, deadline);
+            if (dill_slow(err)) {return -1;}
+            break;
+        case 0x04:
+            // IPv4, already read 1 bytes, read ALEN-1 add + 2 for port
+            err = dill_brecv(obj->u, conn + 5, conn[4] + 1, deadline);
+            if (dill_slow(err)) {return -1;}
+            break;
+        default:
+            errno = EPROTO;
+            return -1;
+    }
+    if (conn[1] == 0x00) {
+        // success - from here on should treat as if attached via tcp_connect
+        obj->connect = 1;
+        return 0;
+    }
+    // failed, it's just a matter of deciding which error code to return
+    switch (conn[1]) {
+        case 0x01:
+            // general SOCKS server failure
+            errno = EIO;
+            break;
+        case 0x02:
+            // connection not allowed by ruleset
+            errno = EACCES;
+            break;
+        case 0x03:
+            // Network unreachable
+            errno = ENETUNREACH;
+            break;
+        case 0x04:
+            // Host unreachable
+            errno = EHOSTUNREACH;
+            break;
+        case 0x05:
+            // Connection Refused
+            errno = ECONNREFUSED;
+            break;
+        case 0x06:
+            // TTL expired
+            errno = ETIMEDOUT;
+            break;
+        case 0x07:
+            // Command not supported
+            errno = EOPNOTSUPP;
+            break;
+        case 0x08:
+            // Address type not supported
+            errno = EAFNOSUPPORT;
+            break;
+        default:
+            // Unknown error code
+            errno = EPROTO;
+    }
+    return -1;
+}
+
+int dill_socks5_close(int s, int64_t deadline) {
```

Comment:
This function doesn't seem to make sense. You either want to close the entire protocol stack (hclose). Or you want to detach the SOCKS5 layer and continue using the underlying protocol. But by quickly skimming over RFC 1928 there seems to be no way to terminate SOCKS5 session in a clean way. If that's so, there's no need for the detach function at all.

Reply:
This is a situation I'm frankly not quite sure how to handle. The way I've implemented it so far, once the SOCKS5 connection is established it simply passes writes/reads to the underlying bytestream (the _bsendl and _brecvl routines feel like a bit of a hack in this regard). 

Another option I considered is similar to how http is handled. The protocol is attached to negotiate the headers and then detached again such that the body is read directly from the underlying bytestream. 

If I combine _attach and _connect, then why not just go one step further and detach all in the same call? 

For TCP outbound: 
    s1 = tcp_connect(<proxy_addr>)
    s2 = socks5_attach_connect_detach(s1, <remote>)
        <internal implementation>
        s1a = socks5_attach(s1)   // create a new socket layer
        socks5_connect(<remote>)
        s2 = socks5_detach(s1a)  // drop back to the underlying socket, i.e. s2 == s1
        </>
    (send/recv to s2 as though you had directly connected to the remote host)

For TCP inbound:
    s1 = tcp_connect(<proxy_addr>)
    s2 = socks5_attach_bind(s1, port)
        <internal implementation>
        s2 = socks5_attach(s1)
        socks5_bind_part1(port) // proxy replies with success on bind
        </>
    s3 = socks5_listen_detach_tcp_accept(s2, &addr)
        <internal implementation>
        socks5_bind_part2(s2, &addr) // proxy replies with remote addr
        s3 = socks5_detach(s2) // here s3 == s1
        </>
    (send/recv to s3 as though you had directly accepted the connection from the remote host)

In either case for TCP, the completion of the interaction with the proxy is you get back the original bytestream socket. How can I name these functions such that it is clear the resulting socket is the underlying socket (so I need to call tcp_close or ssl_detach at the end of the conversation)? 

Have you considered having a generic bytestream_close() function and then let the protocol layers sort out amongst themselves what that means?

Reply:
> If I combine _attach and _connect, then why not just go one step further and detach all in the same call?

Actually, you are right. All you need is two functions, one that intializes the protocol on the client side, other one on the server side. No need for a new socket type or whatnot.

Reply:
My goal here is to have an intuitive interface. It should be clear to the user what kind of socket they have. The result of a successful SOCKS5 CONNECT is a bare TCP connection to the remote server (regardless of the underlying bytestream to the proxy).

Given the following example:
```
s = tcp_connect()
s = tls_attach_client(s)
s = socks5_connect_client(s, &ipaddr)
s = socks5_detach(s)
```

The socket handle I have at the end is for the tls socket, I can bsend/brecv at will. However, the connection from the proxy to the remote server is simply a TCP connection. The traffic would only be encrypted from the client to the proxy. If I want TLS all the way to the remote server then I need to layer another TLS connection inside the first one. 

You also do want to be able to cleanly close the connection. In this case disconnecting the connection cleanly means:
```
s = tls_detach(s)
tcp_close(s)
```

And if I had initiated TLS to the remote endpoint I would end up with tls_detach(), tls_detach(), tcp_close()

Reply:
What about this:

```
s = tcp_connect();
s = tls_attach_client(s);
socks5_client(s, &ipaddr);
// s is still a TLS socket here.
```